server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,351 @@
+        if (value == null) {
+            return 0;
+        }
+        return Integer.valueOf(transform(value));
+    }
+
+    /**
+     * Attempts to remove a value from the keystore. WARNING: Removing keys which have not been added to the keystore
+     * may cause undefined behavior, including future false negatives!!
+     * @param value The value to attempt to remove.
+     * @return true if the value was removed, false otherwise
+     */
+    @Override
+    public boolean remove(Integer value) {
+        if (value == null) {
+            return false;
+        }
+        int transformedValue = transform(value);
+        readLock.lock();
+        try {
+            if (!rbm.contains(transformedValue)) { // saves additional transform() call
+                return false;
+            }
+            // move below into write lock
+            stats.numRemovalAttempts.inc();
+        } finally {
+            readLock.unlock();
+        }
+        writeLock.lock();
+        try {
+            CounterMetric numCollisions = collidedIntCounters.get(transformedValue);
+            if (numCollisions != null) {
+                // This transformed value has had a collision before
+                HashSet<Integer> removalSet = removalSets.get(transformedValue);
+                if (removalSet == null) {
+                    // First time a removal has been attempted for this transformed value
+                    HashSet<Integer> newRemovalSet = new HashSet<>();
+                    newRemovalSet.add(value); // Add the key value, not the transformed value, to the list of attempted removals for this transformedValue
+                    removalSets.put(transformedValue, newRemovalSet);
+                    numCollisions.dec();
+                } else {
+                    if (removalSet.contains(value)) {
+                        return false; // We have already attempted to remove this value. Do nothing
+                    }
+                    removalSet.add(value);
+                    numCollisions.dec();
+                    // If numCollisions has reached zero, we can safely remove all values in removalList
+                    if (numCollisions.count() == 0) {
+                        removeFromRBM(transformedValue);
+                        collidedIntCounters.remove(transformedValue);
+                        removalSets.remove(transformedValue);
+                        return true;
+                    }
+                }
+                return false;
+            }
+            // Otherwise, there's not been a collision for this transformedValue, so we can safely remove
+            removeFromRBM(transformedValue);
+            return true;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    // Helper fn for remove()
+    private void removeFromRBM(int transformedValue) {
+        rbm.remove(transformedValue);
+        stats.size.dec();
+        stats.numSuccessfulRemovals.inc();
+    }
```

Comment:
should we check if the write lock is held here ?
```suggestion
    private void removeFromRBM(int transformedValue) {
        if (!lock.isWriteLockedByCurrentThread()) {
            throw new IllegalStateException("Write Lock must be held when calling this method");
        }
        rbm.remove(transformedValue);
        stats.size.dec();
        stats.numSuccessfulRemovals.inc();
    }
```

Reply:
Good catch